File: mwc/romana/relic/d/lib/libc/string/strncpy.c
/*
 * strncpy.c
 * ANSI 4.11.2.4.
 * Copy one fixed length string to another.
 */

#include <string.h>

char *strncpy(s1, s2, n) char *s1; char *s2; size_t n;
{
	register char *cp;

	for (cp = s1; n; ) {
		n--;
		if ((*cp++ = *s2++) == '\0')
			break;
	}
	/* Pad remainder of s1 with NUL if required. */
	while (n-- > 0)
		*cp++ = '\0';
	return (s1);
}

File: mwc/romana/relic/b/lib/libc/string/i8086/strncpy.m
//////////
/ i8086 C string library.
/ strncpy()
/ ANSI 4.11.2.4.
//////////

//////////
/ char *
/ strncpy(To, From, Count)
/ char *To, *From;
/ int Count;
/
/ Copy up to Count bytes from From to To,
/ then NUL-pad if Count not reached.
//////////

#include <larges.h>

To	=	LEFTARG
From	=	To+DPL
Count	=	From+DPL

	Enter(strncpy_)
	Les	di, To(bp)	/ To address to ES:DI
	mov	cx, Count(bp)	/ Count to CX
	or	cx, cx
	je	2f		/ Copy nothing
	Lds	si, From(bp)	/ From address to DS:SI
	cld

1:	lodsb			/ Fetch From character
	stosb			/ and store through To
	orb	al, al
	loopne	1b		/ Continue if AL!=0 && CX!=0
	rep			/ Finish clearing line
	stosb			/ (does nothing if CX==0)

2:	mov	ax, To(bp)	/ Return the destination
#if	LARGEDATA
	mov	dx, es
#endif
	Leave
